tlatools/src/tlc2/tool/Simulator.java

```diff
@@ -103,41 +114,303 @@ public Simulator(String specFile, String configFile, String traceFile, boolean d
 	private final Action[] impliedActions; // the implied-actions to be checked
 	private final boolean checkDeadlock; // check deadlock?
 	private final boolean checkLiveness; // check liveness?
-	private long numOfGenStates;
+	
+	// The total number of states generated by all workers. Written to concurrently.
+	private AtomicLong numOfGenStates;
+	
 	// private Action[] actionTrace; // SZ: never read locally
 	private final String traceFile;
+
+	// The maximum length of a simulated trace.
 	private final long traceDepth;
+
+	// The maximum number of total traces to generate.
 	private final long traceNum;
+
+	// The number of worker threads to use for simulation.
+	private int numWorkers = 1;
+
 	private final RandomGenerator rng;
 	private final long seed;
 	private long aril;
 	private final ObjLongTable astCounts;
 	private boolean isCancelled; // SZ Feb 24, 2009: cancellation added
 	private Value[] localValues = new Value[4];
 
+	// The set of all initial states for the given spec. This should be only be
+	// computed once and re-used whenever a new random trace is generated. This
+	// variable should not be written to concurrently, but is allowed to be read
+	// concurrently.
+	private StateVec initStates = new StateVec(0);
+
+	// Appended to when a simulation worker completes. We only care about the first
```

Comment:
It might makes sense to support TLC's continue option to keep state exploration going even if an error is found.

Reply:
Here are my thoughts on how the `-continue` flag should work in simulation mode. If `continue=false`, then on any error from a single worker, of any type, we terminate and kill all other workers. For `continue=true`, I think we should consider treating certain error varieties differently. I see two main error categories. 

The first error category indicates that there is likely something wrong with the way the specification is constructed. For example, a runtime evaluation error (e.g. a "type" error). I think that if one worker encounters this type of error, we should kill all workers and terminate, even if `continue=true`, since it's a reasonable indication that other workers might encounter similar issues with the spec. 

The second error category is normal invariant, temporal property violations. Those errors can obviously be "ignored" if `continue=true` i.e. we'll print some information about them but keep all workers running. 

Do you agree with this?

Reply:
Agreed.

By the way, for the first category lets use the term "error (call) stack" (see e.g. tlc2.tool.Tool.setCallStack()) and for the second one simply "error trace".

Reply:
An "error trace" shows the (prefix of a) behavior that violates an invariant or a liveness property. An "error call stack" reports an error in evaluating expressions in the spec or model. The first one means the state space - leading to the error state - has been fully evaluated while the latter means state space exploration stopped when the expression failed. Both conditions should be reported separately.